docs/design/2022-04-15-multi-schema-change.md

```diff
@@ -0,0 +1,126 @@
+
+The Multi-Schema Change DDL jobs have the type `ActionMultiSchemaChange`. In the current worker model, there is a dedicated code path (`onMultiSchemaChange()`) to run these jobs. Only Multi-Schema Change jobs can have sub-jobs.
+
+For example, the DDL
+
+```SQL
+ALTER TABLE t ADD COLUMN b INT, MODIFY COLUMN a CHAR(255);
+```
+
+can be modeled as a job like
+
+```go
+job := &Job {
+    Type: ActionMultiSchemaChange,
+    MultiSchemaInfo: &MultiSchemaInfo {
+        SubJobs: []*SubJob {
+            &SubJob {
+                Type: ActionAddColumn,
+                Args: ...
+            },
+            &SubJob {
+                Type: ActionModifyColumn,
+                Args: ...
+            },
+        }
+    } 
+}
+```
+
+In this way, we pack multiple schema changes into one job. Just like any other jobs, it enters the DDL job queue stored in TiKV and waits for a suitable worker to pick it up.
+
+In normal cases, the worker executes the sub-jobs one by one serially as if they are plain jobs. However, the thing gets complex in abnormal cases.
+
+To ensure the atomicity of a Multi-Schema Change execution, we need to maintain the changing schema object states carefully. Take the above SQL as an example, if the second sub-job `MODIFY COLUMN a CHAR(255)` failed for some reason, the first sub-job should be able to rollback its changes(rollback the added column `b`).
+
+This requirement means we cannot simply public the schema object when a sub-job is finished. Instead, it should stay in a state which is still unnoticeable to users, and wait for the other sub-jobs, finally public all at once when all the sub-jobs are confirmed to be successful. This method is kind of like 2PC: the "commit" can only be started when "prewrites" are complete.
+
+Here is the table of states that should stay for different DDLs. Note the "Next State" means the changes are noticeable to the users.
+
+| DDL (Schema Change)     | Unnoticeable State | Next State |
+|-------------------------|--------------------|------------|
+| Add Column              | Write-Reorg        | Public     |
+| Add Index               | Write-Reorg        | Public     |
+| Drop Column             | Public             | Write-Only |
+| Drop Index              | Public             | Write-Only |
+| Non-reorg Modify Column | Public             | Public     |
+| Reorg Modify Column     | Write-Reorg        | Public     |
+
+To achieve this behavior, we introduce a flag in the sub-job named "non-revertible". This flag is set when a schema object has stepped to the last unnoticeable state. When all sub-jobs become non-revertible, all the related schema objects step to the next state in one transaction. After that, the sub-jobs are executed serially to do the rest.
```

Comment:
Do you need a brief introduction, a non-common job corresponds to a non-revertible state? For example, a "dropIndex" operation changes to this state after it is the public state.

Reply:
I updated the table to make it easier to understand.